### ISSUE TITLE
[enhancement] more flexibility in visClusteringByGroup

### ISSUE THREAD

**Issue by Tixierae** [datastorm-open/visNetwork#71]
Reproducible example:

````
nodes_test = data.frame(id=1:7, label=LETTERS[1:7], group = c(rep("cluster 1",3),rep("cluster 2",4)), group_color = paste("group #", c(1,1,2,2,2,3,3)), stringsAsFactors = FALSE) 

# create complete graph
edges_test = as.data.frame(t(combn(nodes_test[,"id"],2)))
# delete some edges at random
set.seed(1234)
edges_test = edges_test[-sample(1:nrow(edges_test),7),]
colnames(edges_test) = c("from","to")

# assign colors
colors_df = cbind(paste("group #", c(1,2,3)),substring(heat.colors(n=3),1,7))
my_colors = unlist(lapply(nodes_test[,"group_color"], function(x) colors_df[which(x==colors_df[,1]),2]))
nodes_test[,"color"] = my_colors

# clustering by group
legend_nodes = data.frame(label = unique(nodes_test[,"group_color"]), shape = c("ellipse"), color = unique(nodes_test[,"color"]))

vis = visNetwork(nodes_test, edges_test) %>% visLayout(randomSeed = 1234) %>% visLegend(addNodes=legend_nodes,useGroups=FALSE)%>% visClusteringByGroup(groups=names(summary(as.factor(nodes_test[,"group"]))))

vis
````
The first thing is that the two clusters overlap significantly, so that it is tough to differentiate between the two:

![screenshot_clustering](https://cloud.githubusercontent.com/assets/13351315/16005865/df502f6e-3169-11e6-88d4-41538f505ddc.png)

Do you have any fix/suggestion on how to improve that? (note that I really like the fact that the sizes are proportional to the number of nodes within the clusters and would like to keep this nice feature if possible)

Also, would it be possible to:

- change the shape and color of the clusters, so that we see the difference with nodes? (for instance, rectangles instead of circles). For the color, it could be selected automatically as the majority color of the nodes in the cluster (if no majority, randomly)? Thus, the color of the clusters would be somehow consistent with the legend
- remove "cluster on group" from the label? (to save some space, in practice, the names of my clusters are quite long)

Thank you very much in advance

**Comment by bthieurmel** [comment]
Hi,

Don't look at clustering for a long time (and code functions very quickly)...! 

- For the first point, you can try use another solver ``visPhysics(solver = "forceAtlas2Based")``
- Actually, cluster take the shape/color of nodes if all equal, or a default value. So to control that, I now let the possibility to change this default value with ``shape``, ``color`` and ``force`` argument.
- Add also add ``label`` to change ``cluster on group : ``

And I think this is a point (clustering) with a lot of things to do / to improve...

**Comment by Tixierae** [comment]
Thanks for the quick reply! I did not know that I could pass extra arguments. I will try it out now. I will also try to tweak the Physics parameters to improve layout.

**Comment by bthieurmel** [comment]
It's because I add this extra arguments right now.

**Comment by Tixierae** [comment]
Sounds good. Are these improvements already available in the dev version?

**Comment by bthieurmel** [comment]
Yes, of course.

**Comment by Tixierae** [comment]
Awesome, thanks!

**Comment by Tixierae** [comment]
Any suggestion why double-clicking on the clusters would expand them in RStudio viewer but not in Shiny app (seen in browser)?

**Comment by bthieurmel** [comment]
Can't reproduce this bug... I'll able to open cluster with double-clicked in a shiny apps (Mozilla, Google Chrome, Edges, and IE 11).

**Comment by Tixierae** [comment]
and with the following parameters? `visEdges(smooth = FALSE, width = 0.05, color=list(color="grey",highlight="red",hover="red"), selectionWidth=0.5, hoverWidth=0.5) %>% visInteraction(hover = TRUE, hoverConnectedEdges = TRUE)`

**Comment by bthieurmel** [comment]
Always ok for me... Can't you see some error/information message in a web development console ?

**Comment by Tixierae** [comment]
I can't see any warning or error. Everything is fine, it is just that when I double click nothing happens. My edges have titles (so that a label shows up when I hover them)... maybe that's the reason...

**Comment by bthieurmel** [comment]
Ok for me also with ``label``, ``title``, ... Can you send me simple shiny app with the problem ? And perhaps your ``sessionInfo()``

**Comment by Tixierae** [comment]
I found the issue. It has to do with `selectBy` (repro example below)

````
library(visNetwork)

server <- function(input, output) {

nodes_test = data.frame(id=1:7, label=LETTERS[1:7], group = c(rep("cluster 1",3),rep("cluster 2",4)), group_color = paste("group #", c(1,1,2,2,2,3,3)), stringsAsFactors = FALSE) 

# create complete graph
edges_test = as.data.frame(t(combn(nodes_test[,"id"],2)))

# delete some edges at random
set.seed(1234)
edges_test = edges_test[-sample(1:nrow(edges_test),7),]
colnames(edges_test) = c("from","to")

# assign colors
colors_df = cbind(paste("group #", c(1,2,3)),substring(heat.colors(n=3),1,7))
my_colors = unlist(lapply(nodes_test[,"group_color"], function(x) colors_df[which(x==colors_df[,1]),2]))
nodes_test[,"color"] = my_colors

vis = visNetwork(nodes_test, edges_test) %>% visLayout(randomSeed = 1234) 

legend_nodes = data.frame(label = unique(nodes_test[,"group_color"]), shape = c("ellipse"), color = unique(nodes_test[,"color"]))

vis = vis %>% visLegend(addNodes=legend_nodes,useGroups=FALSE)

my_groups = names(summary(as.factor(nodes_test[,"group"])))

vis = vis %>% visClusteringByGroup(groups=my_groups, color="blue", label="", force=TRUE) %>% visPhysics(solver = "forceAtlas2Based")

vis = vis %>% visOptions(selectedBy = "group_color", highlightNearest = list(enabled = TRUE, hover = TRUE), clickToUse = FALSE) # if you comment out this line, the problem disappears

output$"graph" = renderVisNetwork({vis})

}

ui <- fluidPage(
  visNetworkOutput("graph", height="580px")
)

shinyApp(ui = ui, server = server)
````

**Comment by bthieurmel** [comment]
Thanks, got the error now. So not a shiny / browser problem. I'll have a look tomorrow if possible the correct this.

**Comment by Tixierae** [comment]
Thanks!

**Comment by bthieurmel** [comment]
Normally fix on last version (bad interaction between real nodes / cluster nodes & highlight function)

**Comment by Tixierae** [comment]
Awesome! Many thanks for the quick fix!

**Comment by Tixierae** [comment]
To go back to my initial issue, tweaking the physics parameters allows to add some space between the two clusters, but it also impacts the aspect of the nodes inside the clusters. And the optimal parameters when the clusters are collapsed are not the same as when they are expanded (when double-clicking it looks bad).

So, would it be possible to set the cluster parameters independently of the node parameters?

**Comment by bthieurmel** [comment]
Clustering seems to be very "complicated" to set as a global simple configuration, and it's very new for me.

First, we can just set a global physics to the network. 

But, with a lot time, we can try to set clustering better...! I don't have a lot time actually, but I add a little things : 

- if you pass coordinates to your nodes (with ``visIgraphLayout`` for example)
- and if you disable physics
- you can retrieve your positions...

````
visNetwork(nodes_test, edges_test) %>% 
  visLayout(randomSeed = 1234) %>% 
  visLegend(addNodes=legend_nodes,useGroups=FALSE)%>% 
  visIgraphLayout(randomSeed = 123)%>%
  visClusteringByGroup(groups=names(summary(as.factor(nodes_test[,"group"])))) %>%
  visPhysics(enabled = FALSE)
````

In my todo list, but don't know when.

**Comment by Tixierae** [comment]
No worries. Thank you very much for your quick and valuable help. 

In my case, I think that a quick and dirty solution could be to reduce the size of the clusters, and use a shape thinner than the default `"database"`. `"triangle"` is a nice option, but do you know if fontawesome icons are available for the clusters?

I tried:

 `vis = vis %>% visClusteringByGroup(groups=my_groups, label="", shape = "icon", icon = list(code = "f0c0", size = 30, color="3399FF"), force=TRUE) %>% addFontAwesome() %>% visPhysics(solver = "forceAtlas2Based") ` 

like in your example here: http://search.r-project.org/library/visNetwork/html/addFontAwesome.html
but it does not work (icon is not recognized as an argument). I know that in your example you use `visGroups`, but doing so changes the shape of the nodes within the clusters, whereas I want to change the shape of the clusters themselves.

**Comment by bthieurmel** [comment]
No, can't use icons in cluster actually, but perhaps in the futur...

**Comment by Tixierae** [comment]
ok, I need to wait then! Thanks!
